airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java

```diff
@@ -355,17 +361,32 @@ public WebBackendConnectionRead webBackendUpdateConnection(final WebBackendConne
     final List<UUID> operationIds = updateOperations(webBackendConnectionUpdate);
     final ConnectionUpdate connectionUpdate = toConnectionUpdate(webBackendConnectionUpdate, operationIds);
 
-    ConnectionRead connectionRead;
-    final boolean needReset = MoreBooleans.isTruthy(webBackendConnectionUpdate.getWithRefreshedCatalog());
+    final UUID connectionId = webBackendConnectionUpdate.getConnectionId();
+
+    final ConfiguredAirbyteCatalog existingConfiguredCatalog =
+        configRepository.getConfiguredCatalogForConnection(connectionId);
+    final io.airbyte.protocol.models.AirbyteCatalog existingCatalog = CatalogHelpers.configuredCatalogToCatalog(existingConfiguredCatalog);
+    final AirbyteCatalog apiExistingCatalog = CatalogConverter.toApi(existingCatalog);
+    final AirbyteCatalog newAirbyteCatalog = webBackendConnectionUpdate.getSyncCatalog();
+    final CatalogDiff catalogDiff = connectionsHandler.getDiff(apiExistingCatalog, newAirbyteCatalog);
+
+    final List<StreamDescriptor> apiStreamsToReset = getStreamsToReset(catalogDiff);
+    List<io.airbyte.protocol.models.StreamDescriptor> streamsToReset =
+        apiStreamsToReset.stream().map(ProtocolConverters::streamDescriptorToProtocol).toList();
 
+    ConnectionRead connectionRead;
     connectionRead = connectionsHandler.updateConnection(connectionUpdate);
 
-    if (needReset) {
+    if (!streamsToReset.isEmpty()) {
+      final ConnectionIdRequestBody connectionIdRequestBody = new ConnectionIdRequestBody().connectionId(connectionId);
+      final ConnectionStateType stateType = getStateType(connectionIdRequestBody);
+
+      if (stateType == ConnectionStateType.LEGACY || stateType == ConnectionStateType.NOT_SET || stateType == ConnectionStateType.GLOBAL) {
```

Comment:
Also, one more thing I think we discussed is that we probably want to allow the diff result `streamsToReset` to be used when the state type is `NOT_SET`. 

The state type will be `NOT_SET` if the connection is a non-incremental one or if nothing has been synced yet. Whether we use all streams for connection or `streamsToReset` doesn't matter for the resulting `state` (EmptyAirbyteSource will just emit a null state in either case, since there is not current state - code [here](https://github.com/airbytehq/airbyte/blob/374e265fcbdfcef41a77bfcd7ec2acd2450d181d/airbyte-workers/src/main/java/io/airbyte/workers/internal/EmptyAirbyteSource.java#L98-L99)). But using `streamsToReset` vs `getAllStreamsForConnection` _does_ affect what will happen to the data in the destination. And it seems useful to only clear out the data for the streams that are being updated/added, rather than clearing out all data.

@benmoriceau or @andyjih do you agree with the above?

Reply:
If `streamsToReset` only includes the streams that were updated/added, then yes we should only update the relevant streams and not all streams.

Reply:
`streamsToReset` will include all streams that are updated, added, or deleted. But we will only end up clearing out the data for streams that are updated or added in the destination, because those are the only ones that are still in the catalog after it is updated in this endpoint. Streams that aren't in the catalog are not modified in the destination, so deleted streams won't have their destination data affected

Reply:
Wouldn't this be handled correctly by the empty airbyte source reset whether we pass in streamsToReset or all streams in a connection?

Reply:
The EmptyAirbyteSource will act the same way in either case - it will output no state, since none exists; the difference is in this logic where we decide whether to set the destination sync mode of a stream in the catalog to `OVERWRITE` or `APPEND` based on whether the stream is present in `streamsToReset`: https://github.com/airbytehq/airbyte/blob/4e80a67d892033dc625f8f401c82473b5424e38e/airbyte-scheduler/scheduler-persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java#L102-L111

What I'm proposing here is that when the state is `NOT_SET`, we probably only want to set the streams in the diff result to `OVERWRITE`, so that only those streams are cleared out in the destination

Reply:
@benmoriceau I don't think the current sync mode configuration of the connection matters when we are performing a reset, since from the code I linked above you can see that we are choosing which sync mode to use for each stream in the catalog when performing a reset. So what I was saying was that in the case that we are updating a connection whose state is currently `NOT_SET`, then we can just use the results of the diff to decide which streams should be reset, and the above code will set those to `FULL_REFRESH/OVERWRITE`, and have the rest of the streams (the ones without any diff) be set to `FULL_REFRESH/APPEND`. That way, only the streams in the diff will have their data cleared in the destination, and others will be unaffected.

And all of the above can be accomplished by just setting `streamsToReset` to the results of the diff in the case that the current connection state is `NOT_SET`